### ISSUE TITLE
Megaldon crashing at guppy_basecall_server init, but no error message

### ISSUE THREAD

**Issue by CodingKaiser** [nanoporetech/megalodon#171]
Dear all, I am trying to run megalodon on some fast5 files but running into an inexplicable (afaik) error.

```
guppy_path=$(which guppy_basecall_server)

megalodon \
        hp_fast5/ \
        --guppy-params "-d ~/ont/guppy_models/rerio/basecall_models" \
        --guppy-server-path $guppy_path \
        --guppy-config res_dna_r941_min_modbases_5mC_CpG_v001.cfg \
        --overwrite \
        --outputs mods \
        --reference /cluster/work/nme/genomes/ONT/pSpCas9hp.fa \
        --mod-motif m CG 0 \
        --devices cuda:all
```

This command errors out with the usual message:

```
[11:37:00] Running Megalodon version 2.3.3
[11:37:00] Loading guppy basecalling backend
****************************************************************************************************
	ERROR: Guppy server initialization failed. See guppy logs in [--output-directory] for more details.
		Try running the guppy server initialization command found in log.txt in order to pinpoint the source of this issue.
****************************************************************************************************
```

Now comes the problem: Running the command used to invoke the `guppy_basecall_server` in order to diagnose the error results in no errors being thrown at all, with the server seeming to run completely fine.

This is the output I get when I start the server manually using command `/cluster/work/nme/software/apps/guppy_gpu/4.2.2/bin/guppy_basecall_server -p auto -l megalodon_results/guppy_log -c res_dna_r941_min_modbases_5mC_CpG_v001.cfg --post_out --quiet -x cuda:0 -d ~/ont/guppy_models/rerio/basecall_models`:

```
ONT Guppy basecall server software version 5.0.11+2b6dbff, client-server API version 7.0.0
log path:            megalodon_results/guppy_log
chunk size:          2000
chunks per runner:   512
max queued reads:    2000
num basecallers:     4
num socket threads:  2
max returned events: 50000
gpu device:          cuda:all
kernel path:
runners per device:  4

Config loaded:
config file:               /cluster/home/falnoe/ont/guppy_models/rerio/basecall_models/res_dna_r941_min_modbases_5mC_CpG_v001.cfg
model file:                /cluster/home/falnoe/ont/guppy_models/rerio/basecall_models/res_dna_r941_min_modbases_5mC_CpG_v001.jsn
model version id           None
adapter scaler model file: None
Starting server on port: 34492
```

In short, I only get the error when guppy_basecall_server is started via megalodon.

I am using:
Python: 3.7.4
megalodon: 2.3.3
guppy_gpu: 4.2.2

I have tried with megalodon 2.3.1 with no avail as well as other versions of guppy (< 4.2.2) and guppy 5.0.11. I therefore do not believe it to be a guppy issue.

Much appreciated.

**Comment by marcus1487** [comment]
Could you provide the guppy and megalodon log files from the failed run?

**Comment by CodingKaiser** [comment]
Sure thing. Here is the megalodon log (megalodon_results/log.txt) file output:

```
[14:29:02] Running Megalodon version 2.3.3
DBG 14:29:02 : Command: """/cluster/home/falnoe/ont/taiyaki/venv/bin/megalodon hp_fast5/ --guppy-params -d ~/ont/guppy_models/rerio/basecall_models --guppy-server-path /cluster/work/groupname/software/apps/guppy_gpu/5.0.11/bin/guppy_basecall_server --guppy-config res_dna_r941_min_modbases_5mC_CpG_v001.cfg --overwrite --outputs mods --reference /cluster/work/groupname/genomes/ONT/pSpCas9hp.fa --mod-motif m CG 0 --devices cuda:0""" --- MainProcess-MainThread megalodon.py:1750
[14:29:02] Loading guppy basecalling backend
DBG 14:29:02 : Guppy version: "5.0.11" --- MainProcess-MainThread backends.py:842
DBG 14:29:02 : Pyguppy version: "5.0.11" --- MainProcess-MainThread backends.py:843
DBG 14:29:02 : guppy server init command: "/cluster/work/groupname/software/apps/guppy_gpu/5.0.11/bin/guppy_basecall_server -p auto -l megalodon_results/guppy_log -c res_dna_r941_min_modbases_5mC_CpG_v001.cfg --post_out --quiet -x cuda:0 -d ~/ont/guppy_models/rerio/basecall_models" --- MainProcess-MainThread backends.py:919
****************************************************************************************************
        ERROR: Guppy server initialization failed. See guppy logs in [--output-directory] for more details.
                Try running the guppy server initialization command found in log.txt in order to pinpoint the source of this issue.
****************************************************************************************************
```

There aren't any log files inside `megalodon_results/guppy_log/`.

**Comment by marcus1487** [comment]
That is quite odd that there is no guppy log from the megalodon call. It seems that at least your standalone guppy server call should have left a log in that directory. Is it possible that this file was cleaned up at some point?

**Comment by CodingKaiser** [comment]
My apologies. The standalone call does leave behind a log file. Here are the contents of that file (`megalodon_results/guppy_log/guppy_basecall_server_log-2021-07-15_15-57-59.log`)

```
2021-07-15 15:57:59.476242 [guppy/message] ONT Guppy basecall server software version 5.0.11+2b6dbff, client-server API version 7.0.0
log path:            megalodon_results/guppy_log
chunk size:          2000
chunks per runner:   512
max queued reads:    2000
num basecallers:     4
num socket threads:  2
max returned events: 50000
gpu device:          cuda:all
kernel path:
runners per device:  4
2021-07-15 15:57:59.477376 [guppy/info] crashpad_handler not supported on this platform.
2021-07-15 15:57:59.523200 [guppy/info] Listening on port 36374.
2021-07-15 15:58:00.070653 [guppy/info] CUDA device 0 (compute 6.1) initialised, memory limit 11721506816B (11577720832B free)
2021-07-15 15:58:09.058220 [guppy/message]
Config loaded:
config file:               /cluster/home/falnoe/ont/guppy_models/rerio/basecall_models/res_dna_r941_min_modbases_5mC_CpG_v001.cfg
model file:                /cluster/home/falnoe/ont/guppy_models/rerio/basecall_models/res_dna_r941_min_modbases_5mC_CpG_v001.jsn
model version id           None
adapter scaler model file: None
2021-07-15 15:58:09.088406 [guppy/message] Starting server on port: 36374
2021-07-15 15:59:43.966476 [guppy/message] SIGINT received. Terminating server.
2021-07-15 15:59:44.170824 [guppy/message] The basecall server has shut down successfully.
2021-07-15 15:59:45.129077 [guppy/info] Stats for model /cluster/home/falnoe/ont/guppy_models/rerio/basecall_models/res_dna_r941_min_modbases_5mC_CpG_v001.jsn, 4 runners/device, 512 chunks/run, 2000 blocks/chunk, lifetime 101.72 s
CUDA device 0: 0 runs with 0 chunks (-nan%), 0 samples (-nan%), avg max size -nan, avg size -nan (-nan% of max), 0 samples/s
```

**Comment by damioresegun** [comment]
@CodingKaiser did you find a way around this? I'm getting the same initialisation error

**Comment by CodingKaiser** [comment]
I actually just found out what the problem was. For me, the argument `--guppy-params "-d ~/ont/guppy_models/rerio/basecall_models"` was the issue. Replacing the `~` with the absolute path to me home directory so that you get `--guppy-params "-d /cluster/home/falnoe/ont/guppy_models/rerio/basecall_models` fixed the problem for me.

This explains why it works when you run guppy_basecall_server yourself but not when you call it from megalodon: On the command line, `~` is expanded by the shell. However, Python will not expand `~` if passed inside a string unless you specifically call something like `os.path.expanduser` on it.

**Comment by damioresegun** [comment]
Ahh I see, that makes sense. Unfortunately this does not work for because even the guppy_basecall_server initialisation command itself did not work. Thanks for replying though!

**Comment by marcus1487** [comment]
This issue was resolved due to the unexpanded `~` in the command line. Closing this issue.
